rasa/engine/graph.py

```diff
@@ -0,0 +1,196 @@
+from __future__ import annotations
+from abc import ABC, abstractmethod
+from collections import ChainMap
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, List, Optional, Text, Type
+
+from rasa.engine.runner.dask import logger
+import rasa.shared.utils.common
+
+
+@dataclass
+class SchemaNode:
+    """Represents one node in the schema."""
+
+    needs: Dict[Text, Text]
+    uses: Type[GraphComponent]
+    constructor_name: Text
+    fn: Text
+    config: Dict[Text, Any]
+    eager: bool = False
+    is_target: bool = False
+    is_input: bool = False
+    resource_name: Optional[Text] = None  # TODO: type Resource
+
+
+GraphSchema = Dict[Text, SchemaNode]
+
+
+class GraphComponent(ABC):
+    """Interface for any component which will run in a graph."""
+
+    # TODO: This doesn't enforce that it exists in subclasses..
+    default_config: Dict[Text, Any]
+
+    @classmethod
+    @abstractmethod
+    def create(
+        cls, config: Dict, execution_context: ExecutionContext
+    ) -> GraphComponent:
+        """Creates a new graph component.
+
+        Args:
+            config: This config overrides the `default_config`
+            execution_context: Information about the current graph run.
+
+        Returns: An instantiated GraphComponent
+        """
+        ...
+
+    @classmethod
+    def load(cls, config: Dict, execution_context: ExecutionContext) -> GraphComponent:
+        """The load method is for creating a component using persisted data.
+
+        Args:
+            config: This config overrides the `default_config`
+            execution_context: Information about the current graph run.
+
+        Args:
+            config: This config overrides the `default_config`
+            execution_context: Information about the current graph run.
+
+        Returns: An instantiated, loaded GraphComponent
+        """
+        return cls.create(config, execution_context)
+
+    @abstractmethod
+    def supported_languages(self) -> List[Text]:
+        """Determines which languages this component can work with."""
+        ...
+
+    @abstractmethod
+    def required_packages(self) -> List[Text]:
+        """Any extra python dependencies required for this component to run."""
+        ...
+
+
+@dataclass
+class ExecutionContext:
+    """Holds information about a single graph run."""
+
+    graph_schema: GraphSchema = field(repr=False)
+    model_id: Text
+    diagnostic_data: bool = False
```

Comment:
optional: A bit more in line with `is_finetuning`
```suggestion
   should_add_diagnostic_data: bool = False
```